### PR TITLE
WordAds Block: AMP Support

### PR DESCRIPTION
**Pull request by jeffersonrabb**
#### Changes proposed in this Pull Request:

This PR brings AMP compatibility to the Jetpack Ads block. 

A couple of gotchas:

1) ~The `Additional ad placements->Top of each page` option will lead to AMP incompatibility in TwentySeventeen, TwentyFifteen, and TwentyFourteen. This is due to https://github.com/Automattic/jetpack/blob/master/modules/wordads/wordads.php#L400-L402. I plan to address this in a separate PR since this doesn't directly relate to the block.~
2) AMP ads will be blocked if `Settings->Reading->Search Engine Visibility->Discourage search engines from indexing this site` is checked. 
3) Make sure you have no ad blockers enabled (I fell for this one more than once :-) )

#### Testing instructions:

- Spin up a Jurassic Ninja instance with this link: ~https://jurassic.ninja/create/?branch=add/amp-ads-support~ http://jurassic.ninja/create/?jetpack-beta&branch=add/amp-ads-support
- Connect Jetpack, and upgrade to Premium.
- Navigate to `/wp-admin/admin.php?page=jetpack#/traffic`.
- Enable Ads. Switch off `Additional ad placements->Top of each page` (see note above)
- Install and activate the AMP plugin: https://wordpress.org/plugins/amp/
- Go to AMP settings and switch to **Transitional** mode.
- Create a post, add Ads block, publish the post.
- View published post, in both AMP and non-AMP environments. You should see roughly identical output in both.

#### Proposed changelog entry for your changes:

* AMP compatibility for the Jetpack Ads block.

### CODE REVIEW

**Comment by dbspringer** [comment]
One question I have is: there's a few different AMP implementations available for self-hosted sites, does this one target one in particular, or is it more of a generic solution? We may need to consider supporting a handful of the most popular AMP options.

Though I suppose if a different version of AMP is used our `is_amp` checks will just return false.

File: modules/wordads/wordads.php
Comment (dbspringer) [inline modules/wordads/wordads.php]: return self::++$amp_section_id; would be a little cleaner, but not big deal.

**Comment by jeffersonrabb** [comment]
> One question I have is: there's a few different AMP implementations available for self-hosted sites, does this one target one in particular, or is it more of a generic solution? We may need to consider supporting a handful of the most popular AMP options.

By "different AMP implementations" do you mean plugins other than the official one: https://wordpress.org/plugins/amp/ ?

File: modules/wordads/wordads.php
Comment (dbspringer) [inline modules/wordads/wordads.php]: We may want to consider extracting this into a separate `is_amp` function to include support for various versions of AMP solutions self-hosted sites might use.

Unless we're firm on only supporting the Jetpack_AMP_Support version, then nevermind :)

**Review by dbspringer — COMMENTED** [review]
Generally looks good, just a question about which AMP implementations we need to support.

**Comment by dbspringer** [comment]
Looks like there's at least three different implementations w/ > 10,000 active installs. Maybe if there's an "official" one we can just say that's the one we support.

Screen Shot 2019-06-18 at 11 15 09 AM

**Comment by jeffersonrabb** [comment]
> Looks like there's at least three different implementations w/ > 10,000 active installs. Maybe if there's an "official" one we can just say that's the one we support.

Yeah, interesting point.  This check https://github.com/Automattic/jetpack/blob/master/3rd-party/class.jetpack-amp-support.php#L56-L72 is already used pretty widely throughout Jetpack, but I'll leave it to @Automattic/jetpack-crew to comment on whether we might need to broaden support?

File: modules/wordads/wordads.php
Comment (jeffersonrabb) [reply modules/wordads/wordads.php]: Assuming you meant `return self::$amp_section_id++;`, done in https://github.com/Automattic/jetpack/pull/12733/commits/b0fc5def3fd6c884fe8ddb2ed4f876203091716c

File: modules/wordads/wordads.php
Comment (jeffersonrabb) [reply modules/wordads/wordads.php]: As noted in https://github.com/Automattic/jetpack/pull/12733#issuecomment-503255093, let's see what Jetpack folks think?

**Comment by westonruter** [comment]
> Maybe if there's an "official" one we can just say that's the one we support.

There is an official one, and it is the one being directly supported: https://wordpress.org/plugins/amp/

This is what runs on WordPress.com

Other plugins that want to shim support can do so via the filter exposed in `Jetpack_AMP_Support`.

**Comment by westonruter** [comment]
Also, the AMP for WP plugin includes an old version of the official AMP plugin, so the same conditionals should work.

**Comment by dbspringer** [comment]
Regardless of whether or not we ad support for other versions, I think it makes sense to extract the check into a separate function -- which I did :)

File: modules/wordads/wordads.php
Comment (westonruter) [inline modules/wordads/wordads.php]: These should use `esc_attr()` no?

File: modules/wordads/wordads.php
Comment (westonruter) [inline modules/wordads/wordads.php]: Using `uniqid` breaks the AMP plugin's post-processor cache. Better to use https://developer.wordpress.org/reference/functions/wp_unique_id/ or a polyfill.

Comment (westonruter) [reply modules/wordads/wordads.php]: Though I suppose this won't run in AMP anyway since it's after the `return`?

Comment (dbspringer) [reply modules/wordads/wordads.php]: > Though I suppose this won't run in AMP anyway since it's after the return?
That's correct.

File: modules/wordads/wordads.php
Comment (dbspringer) [reply modules/wordads/wordads.php]: Thanks, added.

**Review by dbspringer — DISMISSED** [review]
Okay, I think this update is now a solid enough first release while the other updates noted get worked on.

**Comment by jeffersonrabb** [comment]
> 1. The `Additional ad placements->Top of each page` option will lead to AMP incompatibility in TwentySeventeen, TwentyFifteen, and TwentyFourteen. This is due to https://github.com/Automattic/jetpack/blob/master/modules/wordads/wordads.php#L400-L402. I

This is resolved in https://github.com/Automattic/jetpack/pull/12733/commits/877f0cdf0647e43329b89aa26860a11beb5ca9af. For AMP requests, the unit will be inserted above the content, below the title. After discussing with @dbspringer we decided there wasn't any way to match the non-AMP positioning without using Javascript. This isn't precisely the same as the non-AMP placement, but is similar in spirit.

**Review by jeherve — CHANGES_REQUESTED** [review]
I'm getting the following notice when trying to view a post in an AMP view:

```
PHP Notice:  is_amp_endpoint was called <strong>incorrectly</strong>. is_amp_endpoint() was called before the parse_query hook was called. Please see <a href="https://codex.wordpress.org/Debugging_in_WordPress">Debugging in WordPress</a> for more information. (This message was added in version 0.4.2.) in /var/www/html/wp-includes/functions.php on line 4773
[20-Jun-2019 13:49:11 UTC] PHP Stack trace:
[20-Jun-2019 13:49:11 UTC] PHP   1. {main}() /var/www/html/index.php:0
[20-Jun-2019 13:49:11 UTC] PHP   2. require() /var/www/html/index.php:17
[20-Jun-2019 13:49:11 UTC] PHP   3. require_once() /var/www/html/wp-blog-header.php:13
[20-Jun-2019 13:49:11 UTC] PHP   4. require_once() /var/www/html/wp-load.php:37
[20-Jun-2019 13:49:11 UTC] PHP   5. require_once() /var/www/html/wp-config.php:85
[20-Jun-2019 13:49:11 UTC] PHP   6. do_action() /var/www/html/wp-settings.php:525
[20-Jun-2019 13:49:11 UTC] PHP   7. WP_Hook->do_action() /var/www/html/wp-includes/plugin.php:465
[20-Jun-2019 13:49:11 UTC] PHP   8. WP_Hook->apply_filters() /var/www/html/wp-includes/class-wp-hook.php:310
[20-Jun-2019 13:49:11 UTC] PHP   9. WordAds->init() /var/www/html/wp-includes/class-wp-hook.php:286
[20-Jun-2019 13:49:11 UTC] PHP  10. WordAds->insert_adcode() /var/www/html/wp-content/plugins/jetpack/modules/wordads/wordads.php:156
[20-Jun-2019 13:49:11 UTC] PHP  11. WordAds::is_amp() /var/www/html/wp-content/plugins/jetpack/modules/wordads/wordads.php:235
[20-Jun-2019 13:49:11 UTC] PHP  12. Jetpack_AMP_Support::is_amp_request() /var/www/html/wp-content/plugins/jetpack/modules/wordads/wordads.php:78
[20-Jun-2019 13:49:11 UTC] PHP  13. is_amp_endpoint() /var/www/html/wp-content/plugins/jetpack/3rd-party/class.jetpack-amp-support.php:62
[20-Jun-2019 13:49:11 UTC] PHP  14. _doing_it_wrong() /var/www/html/wp-content/plugins/amp/includes/amp-helper-functions.php:275
[20-Jun-2019 13:49:11 UTC] PHP  15. trigger_error() /var/www/html/wp-includes/functions.php:4773
```

Previous art on that topic: #11195

**Comment by jeffersonrabb** [comment]
@dbspringer any issue with skipping `insert_head_iponweb()` and `insert_head_meta()` in AMP environments? https://github.com/Automattic/jetpack/pull/12733/commits/0e1a7ed084796c2d15be35d90023c1bdae9994a4

**Comment by dbspringer** [comment]
Yeah, totally fine 👍

File: modules/wordads/wordads.php
Comment (westonruter) [inline modules/wordads/wordads.php]: To eliminate the `_doing_it_wrong` error, this cannot be called here since it  is before the `wp` action fires. So what you need to do here is move the `is_amp` check _inside_ the `the_content` filter, because that happens after `wp`. An easy thing to fix this would be just to defer the `init` method to the `wp` action, rather then where it is being called now.

Comment (westonruter) [reply modules/wordads/wordads.php]: In other words, try:

```diff
--- a/modules/wordads/wordads.php
+++ b/modules/wordads/wordads.php
@@ -106,7 +106,7 @@ class WordAds {
 	 * @since 4.5.0
 	 */
 	function __construct() {
-		add_action( 'init', array( $this, 'init' ) );
+		add_action( 'wp', array( $this, 'init' ) );
 	}
 
 	/**
```

Comment (jeffersonrabb) [reply modules/wordads/wordads.php]: @dereksmart do you foresee any issues for non-AMP output if I change the action from `init` to `wp`?

Comment (dbspringer) [reply modules/wordads/wordads.php]: 🤔 That's a good question, I don't _think_ it'll be a problem, but I haven't done any testing to verify. You'll have to work with @rclations and @cphilleo to make sure everything is working as intended on this one.

**Comment by jeherve** [comment]
Rebasing to fix tests.

File: modules/wordads/wordads.php
Comment (rclations) [reply modules/wordads/wordads.php]: Tested a sample site, and it doesn't appear to effect ad rendering or ads.txt output - should be g2g 👍

**Review by rclachance — DISMISSED** [review]
w/r/t Ads placements, this looks great & tests well.
Approving pending the change of the init action to run on the `wp` hook instead of the current `init`.

File: modules/wordads/wordads.php
Comment (jeffersonrabb) [reply modules/wordads/wordads.php]: Change made in https://github.com/Automattic/jetpack/pull/12733/commits/65ae218f9ad439dc43fc8a3fa3dd9acabfe110c9

**Comment by jeffersonrabb** [comment]
Rebased.

**Review by jeherve — CHANGES_REQUESTED** [review]
I am now getting the following errors when editing a post that contains Ad blocks:

```
Notice:  Trying to get property 'blog_id' of non-object in wp-content/plugins/jetpack-dev/extensions/blocks/wordads/wordads.php on line 93
Notice:  Trying to get property 'cloudflare' of non-object in wp-content/plugins/jetpack-dev/modules/wordads/wordads.php on line 566
Notice:  Trying to get property 'blog_id' of non-object in wp-content/plugins/jetpack-dev/extensions/blocks/wordads/wordads.php on line 93
Notice:  Trying to get property 'cloudflare' of non-object in wp-content/plugins/jetpack-dev/modules/wordads/wordads.php on line 566
Notice:  Trying to get property 'blog_id' of non-object in wp-content/plugins/jetpack-dev/extensions/blocks/wordads/wordads.php on line 93
Notice:  Trying to get property 'cloudflare' of non-object in wp-content/plugins/jetpack-dev/modules/wordads/wordads.php on line 566
Warning:  Cannot modify header information - headers already sent by (output started at wp-content/plugins/jetpack-dev/extensions/blocks/wordads/wordads.php:93) in wp-includes/option.php on line 947
Warning:  Cannot modify header information - headers already sent by (output started at wp-content/plugins/jetpack-dev/extensions/blocks/wordads/wordads.php:93) in wp-includes/option.php on line 948
```
`$wordads->params` appears to be `null`.

This can even cause a Fatal error if you hide one of the ads on mobile using the block option:

```
Fatal error: Uncaught Error: Call to a member function is_mobile() on null
in wp-content/plugins/jetpack-dev/extensions/blocks/wordads/wordads.php on line 83
```
This does not appear on Jetpack Stable.

**Comment by jeffersonrabb** [comment]
Hmmm, this is tricky. Looks like the new Block fatals came about because of the change in the init action from `init` to `wp` in https://github.com/Automattic/jetpack/pull/12733/commits/65ae218f9ad439dc43fc8a3fa3dd9acabfe110c9. @westonruter, given this change was made to address the AMP `_doing_it_wrong` error (https://github.com/Automattic/jetpack/pull/12733#discussion_r295891779) do you have any suggestions for how to reconcile these two situations?

**Comment by westonruter** [comment]
Why do the fatals occur due to moving the logic from `init` to `wp`, an action that happens later in  WordPress's execution?

**Comment by jeffersonrabb** [comment]
After discussing with @westonruter https://github.com/Automattic/jetpack/pull/12733/commits/2349a05f065f1393e95a122caf90c8e7dffd5429 should resolve the issue by using different initialization hooks for API and non-API requests.

**Comment by jeherve** [comment]
For the most part it seems to work well now. I do, however, run into Fatals from AMP now:

```
[06-Aug-2019 17:04:24 UTC] PHP Fatal error:  Unknown: Cannot use output buffering in output buffering display handlers in Unknown on line 0
[06-Aug-2019 17:04:24 UTC] PHP Stack trace:
[06-Aug-2019 17:04:24 UTC] PHP   1. shutdown_action_hook() /var/www/html/wp-includes/load.php:0
[06-Aug-2019 17:04:24 UTC] PHP   2. do_action() /var/www/html/wp-includes/load.php:954
[06-Aug-2019 17:04:24 UTC] PHP   3. WP_Hook->do_action() /var/www/html/wp-includes/plugin.php:465
[06-Aug-2019 17:04:24 UTC] PHP   4. WP_Hook->apply_filters() /var/www/html/wp-includes/class-wp-hook.php:310
[06-Aug-2019 17:04:24 UTC] PHP   5. wp_ob_end_flush_all() /var/www/html/wp-includes/class-wp-hook.php:286
[06-Aug-2019 17:04:24 UTC] PHP   6. ob_end_flush() /var/www/html/wp-includes/functions.php:4339
[06-Aug-2019 17:04:24 UTC] PHP   7. AMP_Theme_Support::finish_output_buffering() /var/www/html/wp-includes/functions.php:4339
[06-Aug-2019 17:04:25 UTC] PHP   8. AMP_Theme_Support::prepare_response() /var/www/html/wp-content/plugins/amp/includes/class-amp-theme-support.php:1766
[06-Aug-2019 17:04:25 UTC] PHP   9. AMP_Content_Sanitizer::sanitize_document() /var/www/html/wp-content/plugins/amp/includes/class-amp-theme-support.php:2093
[06-Aug-2019 17:04:25 UTC] PHP  10. AMP_Style_Sanitizer->sanitize() /var/www/html/wp-content/plugins/amp/includes/templates/class-amp-content-sanitizer.php:117
[06-Aug-2019 17:04:25 UTC] PHP  11. AMP_Style_Sanitizer->process_link_element() /var/www/html/wp-content/plugins/amp/includes/sanitizers/class-amp-style-sanitizer.php:731
[06-Aug-2019 17:04:25 UTC] PHP  12. AMP_Style_Sanitizer->process_stylesheet() /var/www/html/wp-content/plugins/amp/includes/sanitizers/class-amp-style-sanitizer.php:1186
[06-Aug-2019 17:04:25 UTC] PHP  13. AMP_Style_Sanitizer->prepare_stylesheet() /var/www/html/wp-content/plugins/amp/includes/sanitizers/class-amp-style-sanitizer.php:1315
[06-Aug-2019 17:04:25 UTC] PHP  14. AMP_Style_Sanitizer->parse_stylesheet() /var/www/html/wp-content/plugins/amp/includes/sanitizers/class-amp-style-sanitizer.php:1565
[06-Aug-2019 17:04:25 UTC] PHP  15. Sabberworm\CSS\Parser->parse() /var/www/html/wp-content/plugins/amp/includes/sanitizers/class-amp-style-sanitizer.php:1480
[06-Aug-2019 17:04:25 UTC] PHP  16. Sabberworm\CSS\CSSList\Document::parse() /var/www/html/wp-content/plugins/amp/vendor/sabberworm/php-css-parser/lib/Sabberworm/CSS/Parser.php:38
[06-Aug-2019 17:04:25 UTC] PHP  17. Sabberworm\CSS\CSSList\CSSList::parseList() /var/www/html/wp-content/plugins/amp/vendor/sabberworm/php-css-parser/lib/Sabberworm/CSS/CSSList/Document.php:21
[06-Aug-2019 17:04:25 UTC] PHP  18. Sabberworm\CSS\CSSList\CSSList::parseListItem() /var/www/html/wp-content/plugins/amp/vendor/sabberworm/php-css-parser/lib/Sabberworm/CSS/CSSList/CSSList.php:49
[06-Aug-2019 17:04:25 UTC] PHP  19. Sabberworm\CSS\RuleSet\DeclarationBlock::parse() /var/www/html/wp-content/plugins/amp/vendor/sabberworm/php-css-parser/lib/Sabberworm/CSS/CSSList/CSSList.php:98
[06-Aug-2019 17:04:25 UTC] PHP  20. Sabberworm\CSS\Parsing\ParserState->consumeUntil() /var/www/html/wp-content/plugins/amp/vendor/sabberworm/php-css-parser/lib/Sabberworm/CSS/RuleSet/DeclarationBlock.php:31
[06-Aug-2019 17:04:25 UTC] PHP  21. Sabberworm\CSS\Parsing\ParserState->consumeComment() /var/www/html/wp-content/plugins/amp/vendor/sabberworm/php-css-parser/lib/Sabberworm/CSS/Parsing/ParserState.php:225
```

@westonruter Does this ring a bell?

**Comment by westonruter** [comment]
@jeherve Humm. Are you getting any other error in the log? 

What is the `MEMORY_LIMIT` in your install? I've been able to reproduce that problem if I lower my `MEMORY_LIMIT`: https://github.com/ampproject/amp-wp/issues/1915#issuecomment-480011301

When testing with @jeffersonrabb we didn't see this issue.

**Comment by jeffersonrabb** [comment]
> What is the `MEMORY_LIMIT` in your install?

If this is a memory issue, wouldn't we expect to see the same on Jetpack stable?

**Comment by jeherve** [comment]
> What is the `MEMORY_LIMIT` in your install?

64M, but I see that I have the same problem with `master`, so whatever is happening on my site is not caused by this branch.

I'll let someone else jump in and test this instead of me.

**Review by jeherve — APPROVED** [review]
This tests well for me today. Merging.

**Pull request merged**
